pkg/config/initoptions.go

```diff
@@ -142,19 +181,31 @@ func calcPackageNamespace(packageDir string) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	// Return the first non-empty namespace found. Cluster-scoped
+	// Return the non-empty unique namespace if found. Cluster-scoped
 	// resources do not have namespace set.
+	currentNamespace := metav1.NamespaceNone
 	for _, node := range nodes {
 		rm, err := node.GetMeta()
-		if err != nil {
+		if err != nil || len(rm.ObjectMeta.Namespace) == 0 && !isBuiltinClusterScopedResource(rm) {
+			return "", errors.Errorf(
+				"resource does not have namespace set, a namespace is required to create the resource " +
+					"used for keeping track of past apply operations. Please specify ---inv-namespace.")
+		}
+		// skip builtin cluster scoped resources
+		if len(rm.ObjectMeta.Namespace) == 0 && isBuiltinClusterScopedResource(rm) {
```

Comment:
Looking at this now, I actually think we don't need to keep a list of builtin clusterscoped resources. I think we can use the following rules here:
 * If a resource doesn't have the namespace set, it is either a clusterscoped resource or a namespaced resource without a namespace. In both situations, I think it is fine to require the user to specify the namespace. 
 * If a resource have a namespace set, we assume that the resource is namespace-scoped. 

@seans3 What do you think?

Reply:
Removed.